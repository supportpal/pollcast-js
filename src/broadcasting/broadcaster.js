import { extend } from '../util/helpers'
import Channel from './channel'
import Request from '../http/request'

class Broadcaster {
    #options
    #channels = {}
    #timer = null
    #connected = false
    #events = {}
    #defaults = {
      routes: {
          connect: '',
          receive: ''
      },
      polling: 5,
      token: null
    }

    constructor (options) {
      this.#options = extend(this.#defaults, options)
    }

    on (event, callback) {
      if (typeof this.#events[event] === 'undefined') {
        this.#events[event] = []
      }

      this.#events[event].push(callback)

      return this
    }

    subscribe (channel) {
      const $channel = new Channel({ channel: channel })
      if (typeof this.#channels[channel] === 'undefined') {
        this.#channels[channel] = []
      }

      this.#channels[channel].push($channel)

      return $channel
    }

    fire (event, data) {
      if (typeof this.#events[event] === 'undefined') {
        this.#events[event] = []
      }

      for (const callback in this.#events[event]) {
        if (Object.prototype.hasOwnProperty.call(this.#events[event], callback)) {
          this.#events[event][callback](data)
        }
      }
    }

    connect () {
      if (this.#connected) {
        return
      }

      const self = this

      const request = new Request('POST', this.#options.routes.connect)
      request
        .success(function (xhr) {
          const response = JSON.parse(xhr.responseText)
          if (response.status !== 'success') {
            return
          }

          self.#connected = true
          self.setTime(response.time)
          self.fire('connect', self)

          self.#poll()
        })
        .send({_token: this.#options.token})

      return this
    }

    disconnect () {
      this.#connected = false
      clearTimeout(this.#timer)
      this.#timer = null
      this.fire('disconnect', this)

      return this
    }

    setTime (time) {
      this.#options.time = time
    }

    #poll () {
      const self = this

      const request = new Request('POST', this.#options.routes.receive)
      request
        .success(function (xhr) {
          self.#parseResponse(xhr.responseText)
          self.#timer = setTimeout(self.#poll, (self.#options.polling * 1000))
        })
        .send({
          time: this.#options.time,
          channels: this.#channelData(),
          _token: this.#options.token
        })

      return this
    }

    #channelData () {
      const channelData = {}
      for (const channel in this.#channels) {
        if (!Object.prototype.hasOwnProperty.call(this.#channels, channel)) {
          continue
        }

        if (typeof channelData[channel] === 'undefined') {
          channelData[channel] = []
        }

        for (let i = 0; i < this.#channels[channel].length; i++) {
          const obj = this.#channels[channel][i]
          const events = obj.events
          for (const key in events) {
            channelData[channel].push(key)
          }
        }
      }

      return channelData
    }

    #parseResponse (response) {
      response = JSON.parse(response)
      if (response.status !== 'success') {
        return
      }

      this.setTime(response.time)

      for (const event in response.events) {
        if (!Object.prototype.hasOwnProperty.call(response.events, event)) {
          continue
        }

        const item = response.events[event]
        for (let i = 0; i < item.channels.length; ++i) {
          const channel = item.channels[i]

          for (let c = 0; c < this.#channels[channel].length; ++c) {
            this.#channels[channel][c].fire(item)
          }
        }
      }
    }
}

export default Broadcaster
