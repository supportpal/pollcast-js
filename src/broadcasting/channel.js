import { extend } from '../util/helpers'

class Channel {
    #options
    #listeners = {}
    #defaults = {
      channel: null
    }

    constructor (options) {
      this.#options = extend(this.#defaults, options)
    }

    get listeners() {
        return this.#listeners
    }

    on (event, callback) {
      if (typeof this.#listeners[event] === 'undefined') {
        this.#listeners[event] = []
      }

      this.#listeners[event].push(callback)

      return this
    }

    fire (event) {
      for (const e in this.#listeners) {
        if (!Object.prototype.hasOwnProperty.call(this.#listeners, e) || e !== event.event) {
          continue
        }

        const func = this.#listeners[e]
        if (event.delay !== 0) {
          setTimeout(() => { func[0](event.payload, event) }, (event.delay * 1000))
        } else {
          func[0](event.payload, event)
        }
      }
    }
}

export default Channel
