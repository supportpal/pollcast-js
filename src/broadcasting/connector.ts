import { Connector as BaseConnector } from 'laravel-echo'
import { Channel } from './channels/channel'
import { PrivateChannel } from './channels/private-channel'
import { PresenceChannel } from './channels/presence-channel'
import { Socket } from '../http/socket'

export class Connector extends BaseConnector {
  /**
   * The socket instance.
   */
  socket: Socket | undefined

  /**
   * All of the subscribed channel names.
   */
  channels: { [name: string]: Channel } = {}

  /**
   * Merge the custom options with the defaults.
   */
  protected setOptions (options: any): any {
    super.setOptions(options)

    this.options = Object.assign({
      routes: {
        connect: '',
        receive: '',
        publish: '',
        subscribe: '',
        unsubscribe: ''
      },
      polling: 5000
    }, this.options)

    return options
  }

  /**
   * Create a fresh Socket.io connection.
   */
  connect (): void {
    this.socket = new Socket(this.options, this.csrfToken())
    this.socket.connect()
  }

  /**
   * Get a channel instance by name.
   */
  channel (name: string): Channel {
    if (!this.channels[name]) {
      this.channels[name] = new Channel(this.socket, name, this.options)
    }

    return this.channels[name]
  }

  /**
   * Get a private channel instance by name.
   */
  privateChannel (name: string): PrivateChannel {
    if (!this.channels['private-' + name]) {
      this.channels['private-' + name] = new PrivateChannel(this.socket, 'private-' + name, this.options)
    }

    return this.channels['private-' + name] as PrivateChannel
  }

  /**
   * Get a presence channel instance by name.
   */
  presenceChannel (name: string): PresenceChannel {
    if (!this.channels['presence-' + name]) {
      this.channels['presence-' + name] = new PresenceChannel(
        this.socket,
        'presence-' + name,
        this.options
      )
    }

    return this.channels['presence-' + name] as PresenceChannel
  }

  /**
   * Leave the given channel, as well as its private and presence variants.
   */
  leave (name: string): void {
    const channels = [name, 'private-' + name, 'presence-' + name]

    channels.forEach((name) => {
      this.leaveChannel(name)
    })
  }

  /**
   * Leave the given channel.
   */
  leaveChannel (name: string): void {
    if (this.channels[name]) {
      this.channels[name].unsubscribe()

      delete this.channels[name]
    }
  }

  /**
   * Get the socket ID for the connection.
   */
  socketId (): string {
    return this.socket ? this.socket.id : ''
  }

  /**
   * Disconnect connection.
   */
  disconnect (): void {
    if (this.socket) {
      this.socket.disconnect()
    }
  }
}
