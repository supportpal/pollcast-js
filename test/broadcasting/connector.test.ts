import { Connector } from '../../src/broadcasting/connector'

const mockConnect = jest.fn()
const mockSubscribe = jest.fn()
const mockUnsubscribe = jest.fn()
const mockDisconnect = jest.fn()
jest.mock('../../src/http/socket', () => {
  return {
    Socket: jest.fn().mockImplementation(() => {
      return {
        connect: mockConnect,
        subscribe: mockSubscribe,
        unsubscribe: mockUnsubscribe,
        disconnect: mockDisconnect
      }
    })
  }
})

beforeEach(() => jest.clearAllMocks())

describe('connector', () => {
  it('can construct', () => {
    const connector = new Connector({})
    expect(connector).toBeInstanceOf(Connector)
  })

  it('merges default options', () => {
    const connector = new Connector({ polling: 10000 })
    expect(connector.options.polling).toEqual(10000)
  })

  it('constructor calls connect', () => {
    const connector = new Connector({})
    expect(connector).toBeInstanceOf(Connector)
    expect(mockConnect).toBeCalledTimes(1)
  })

  it('can join new channel', () => {
    const connector = new Connector({})
    connector.channel('foo')
    expect(mockSubscribe).toBeCalledTimes(1)
    expect(connector.channels).toMatchSnapshot()
  })

  it('doesnt join again if already in channel', () => {
    const connector = new Connector({})
    const channel1 = connector.channel('foo')
    const channel2 = connector.channel('foo')
    expect(channel1).toBe(channel2)
    expect(mockSubscribe).toBeCalledTimes(1)
  })

  it('can join private channel', () => {
    const connector = new Connector({})
    connector.privateChannel('foo')
    expect(mockSubscribe).toBeCalledTimes(1)
    expect(connector.channels).toMatchSnapshot()
  })

  it('doesnt join again if already in private channel', () => {
    const connector = new Connector({})
    const channel1 = connector.privateChannel('foo')
    const channel2 = connector.privateChannel('foo')
    expect(channel1).toBe(channel2)
    expect(mockSubscribe).toBeCalledTimes(1)
  })

  it('can join presence channel', () => {
    const connector = new Connector({})
    connector.presenceChannel('foo')
    expect(mockSubscribe).toBeCalledTimes(1)
    expect(connector.channels).toMatchSnapshot()
  })

  it('doesnt join again if already in presence channel', () => {
    const connector = new Connector({})
    const channel1 = connector.presenceChannel('foo')
    const channel2 = connector.presenceChannel('foo')
    expect(channel1).toBe(channel2)
    expect(mockSubscribe).toBeCalledTimes(1)
  })

  it('leaves all channels', () => {
    const connector = new Connector({})
    connector.presenceChannel('foo')
    connector.privateChannel('foo')
    connector.channel('foo')

    expect(Object.keys(connector.channels)).toHaveLength(3)

    connector.leave('foo')

    expect(mockUnsubscribe).toBeCalledTimes(3)
    expect(Object.keys(connector.channels)).toHaveLength(0)
  })

  it('leaves specified channel', () => {
    const connector = new Connector({})
    connector.channel('foo')

    expect(Object.keys(connector.channels)).toHaveLength(1)

    connector.leaveChannel('foo')

    expect(mockUnsubscribe).toBeCalledTimes(1)
    expect(Object.keys(connector.channels)).toHaveLength(0)
  })

  it('leave does nothing if specified channel doesnt exist', () => {
    const connector = new Connector({})

    connector.leaveChannel('foo')

    expect(mockUnsubscribe).toBeCalledTimes(0)
  })

  it('returns socket id', () => {
    const connector = new Connector({})
    expect(connector.socketId()).toEqual(undefined)
  })

  it('returns empty socket id when socket is not defined', () => {
    const connector = new Connector({})
    connector.socket = undefined
    expect(connector.socketId()).toEqual('')
  })

  it('can disconnect', () => {
    const connector = new Connector({})
    connector.disconnect()
    expect(mockDisconnect).toBeCalledTimes(1)
  })

  it('disconnect does nothing if socket is not set', () => {
    const connector = new Connector({})
    connector.socket = undefined
    connector.disconnect()
    expect(mockDisconnect).toBeCalledTimes(0)
  })
})
