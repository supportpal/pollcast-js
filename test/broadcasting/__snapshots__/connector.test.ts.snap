// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`connector can join new channel 1`] = `
Object {
  "foo": Channel {
    "eventFormatter": EventFormatter {
      "namespace": "App.Events",
    },
    "name": "foo",
    "options": Object {
      "auth": Object {
        "headers": Object {},
      },
      "authEndpoint": "/broadcasting/auth",
      "broadcaster": "pusher",
      "csrfToken": null,
      "host": null,
      "key": null,
      "namespace": "App.Events",
      "polling": 5000,
      "routes": Object {
        "connect": "",
        "publish": "",
        "receive": "",
        "subscribe": "",
        "unsubscribe": "",
      },
    },
    "socket": Object {
      "connect": [MockFunction] {
        "calls": Array [
          Array [],
        ],
        "results": Array [
          Object {
            "type": "return",
            "value": undefined,
          },
        ],
      },
      "disconnect": [MockFunction],
      "subscribe": [MockFunction] {
        "calls": Array [
          Array [
            "foo",
          ],
        ],
        "results": Array [
          Object {
            "type": "return",
            "value": undefined,
          },
        ],
      },
      "unsubscribe": [MockFunction],
    },
  },
}
`;

exports[`connector can join presence channel 1`] = `
Object {
  "presence-foo": PresenceChannel {
    "eventFormatter": EventFormatter {
      "namespace": "App.Events",
    },
    "name": "presence-foo",
    "options": Object {
      "auth": Object {
        "headers": Object {},
      },
      "authEndpoint": "/broadcasting/auth",
      "broadcaster": "pusher",
      "csrfToken": null,
      "host": null,
      "key": null,
      "namespace": "App.Events",
      "polling": 5000,
      "routes": Object {
        "connect": "",
        "publish": "",
        "receive": "",
        "subscribe": "",
        "unsubscribe": "",
      },
    },
    "socket": Object {
      "connect": [MockFunction] {
        "calls": Array [
          Array [],
        ],
        "results": Array [
          Object {
            "type": "return",
            "value": undefined,
          },
        ],
      },
      "disconnect": [MockFunction],
      "subscribe": [MockFunction] {
        "calls": Array [
          Array [
            "presence-foo",
          ],
        ],
        "results": Array [
          Object {
            "type": "return",
            "value": undefined,
          },
        ],
      },
      "unsubscribe": [MockFunction],
    },
  },
}
`;

exports[`connector can join private channel 1`] = `
Object {
  "private-foo": PrivateChannel {
    "eventFormatter": EventFormatter {
      "namespace": "App.Events",
    },
    "name": "private-foo",
    "options": Object {
      "auth": Object {
        "headers": Object {},
      },
      "authEndpoint": "/broadcasting/auth",
      "broadcaster": "pusher",
      "csrfToken": null,
      "host": null,
      "key": null,
      "namespace": "App.Events",
      "polling": 5000,
      "routes": Object {
        "connect": "",
        "publish": "",
        "receive": "",
        "subscribe": "",
        "unsubscribe": "",
      },
    },
    "socket": Object {
      "connect": [MockFunction] {
        "calls": Array [
          Array [],
        ],
        "results": Array [
          Object {
            "type": "return",
            "value": undefined,
          },
        ],
      },
      "disconnect": [MockFunction],
      "subscribe": [MockFunction] {
        "calls": Array [
          Array [
            "private-foo",
          ],
        ],
        "results": Array [
          Object {
            "type": "return",
            "value": undefined,
          },
        ],
      },
      "unsubscribe": [MockFunction],
    },
  },
}
`;
